src/server/frontend_wayland/session_lock_v1.cpp

```diff
@@ -59,10 +59,23 @@ class SessionLockV1 : wayland::SessionLockV1
     ~SessionLockV1();
 
 private:
+    class SessionLockV1Observer : public mir::scene::SessionLockObserver
+    {
+    public:
+        explicit SessionLockV1Observer (mf::SessionLockV1& lock);
+        void on_lock() override;
+        void on_unlock() override;
+
+    private:
+        SessionLockV1& lock;
+    };
+
     void get_lock_surface(wl_resource* id, wl_resource* surface, wl_resource* output) override;
+    void destroy() override;
+    void unlock_and_destroy() override;
 
     mf::SessionLockManagerV1& manager;
-    std::unique_ptr<frontend::ScreenLockHandle> const lock_handle;
+    std::shared_ptr<SessionLockV1Observer> const observer;
```

Comment:
```suggestion
    /* Lifetime annotation:
     * This is safe even though `SessionLockV1Observer` keeps a reference to `this`
     * because the `shared_ptr` is only only shared with the `ObserverRegistrar<>`
     * (which uses a `weak_ptr` internally) and observations are delegated to the
     * `WaylandExecutor` so all interactions are single-threaded
     */
    std::shared_ptr<SessionLockV1Observer> const observer;
```